Add a RangeSpecificDiagnosticConsumer

jrose-apple opened the pull request:
If a top-level diagnostic is in a particular source range, the RangeSpecificDiagnosticConsumer will funnel it and any attached notes to a particular "sub-consumer" designated for that range (intended to be used with whole files). If it's not in a range designated for any sub-consumer, the diagnostic is passed to all registered sub-consumers.

This is intended to be used for batch mode, so that diagnostics that are definitely associated with a particular file can be emitted in that file's .dia output, while diagnostics that may be associated with other files (such as those that come from Clang) will still get presented to the user.

jrose-apple commented:
@swift-ci Please smoke test

davidungar commented on lib/AST/DiagnosticConsumer.cpp:
```diff
@@ -22,7 +22,72 @@
+
+FileSpecificDiagnosticConsumer::FileSpecificDiagnosticConsumer(
+    MutableArrayRef<ConsumerPair> consumers) {
+  for (ConsumerPair &pair : consumers) {
```
My style would be to extract out the body of this constructor into a function called something like `buildSortedConsumerMap`. That function would in turn call `populateMap` then `sortMap`. Even though the code is clear as is, I like to make it even easier to read quickly.

jrose-apple replied:
I don't think this one pulls its weight. I added comments, though.

davidungar commented on include/swift/AST/DiagnosticConsumer.h:
```diff
@@ -111,6 +112,35 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
+  SmallVector<std::unique_ptr<DiagnosticConsumer>, 4> SubConsumers;
+  SmallVector<std::pair<CharSourceRange, unsigned>, 4> LocationToConsumerMap;
+  static const unsigned NO_SPECIFIC_CONSUMER = -1;
+  unsigned CurrentGroupConsumer = NO_SPECIFIC_CONSUMER;
```
Would this work as a local variable inside of `handleDiagnostic`? I don't see other uses of it, and AFAICT it is set each time that routine is called.

davidungar replied:
Ah--now I see it is set for non-notes and reused for notes. A better name would be `GroupConsumerForPreviousNonNote`. No, that's not quite right. At least a Doxygen comment to explain it would help.

davidungar replied:
Glad you like the idea. Tx.

davidungar commented on lib/AST/DiagnosticConsumer.cpp:
```diff
@@ -22,7 +22,72 @@
+    StringRef FormatString, ArrayRef<DiagnosticArgument> FormatArgs,
+    const DiagnosticInfo &Info) {
+  if (Kind != DiagnosticKind::Note) {
+    // When emitting a new top-level diagnostic, figure out if it belongs to
```
As I get further into this code, I would like `handleDiagnostic` to be something like this:
[There would be a static function in this class called `shouldResetCurrentGroupConsumer` that would take a diagnostic as a parameter and return true iff not a note.]
This function would only say:
    if (shouldResetCurrentGroupConsumer(Kind)) {
      CurrentGroupConsumer = findConsumerFor(Loc);
  }
  consumeDiagnostic(CurrentGroupConsumer, SM, Loc, Kind, FormatString, FormatArgs, Info)

The idea is to more clearly manifest the two-step nature of the process.

Another option would be to have a function that returned a vector consisting of either all the SubConsumers or just the one appropriate SubConsumer for that diagnostic. That would eliminate the if-else with the duplicated call to handleDiagnostic. Or have a `forAllRelevantConsumers` function that took a closure and either replicated the diagnostic handling or not.

I guess I'm hoping to break this function up a bit to make it easier for eyes like mine to see the structure of the computation.

davidungar replied:
Glad you like the idea. Tx.

davidungar reviewed (commented):
Looks very clever and elegant. When I read the code I had to stare at it for a bit to see what was going on, as my comments show. I'd be tempted to break it down into small pieces (small functions) to more directly manifest the structure of the computation.

davidungar commented on unittests/AST/DiagnosticConsumerTests.cpp:
```diff
@@ -0,0 +1,361 @@
+  topConsumer.finishProcessing();
+}
+
+TEST(FileSpecificDiagnosticConsumer, CrossFileNotes) {
```
I find these test routines a bit daunting. Would it make sense to add comments and/or break them up a bit? I'm thinking of a comment or subroutine name like: "Ensure that a Note ends up in the same consumer as the most recent non-note." Or some such, so a reader can gather the intention without delving into the code.

Another approach: list the ways you can imagine it failing, and test each one. That's probably what this is, but it's hard to tell at a glance. 

At a glance, there might be a factoring that could make these test routines a lot more concise. Just a thought...

jrose-apple replied:
I admit to testing multiple things in each function, since the setup is so clumsy. But ownership and having all the expectations up front makes it hard to make the setup *not* clumsy. Maybe making the diagnostic *emission* less noisy would help, so you could compare the expected per-file output with what's going to be emitted?

jrose-apple replied:
No, it turns out I can come up with reasonable names for all these tests.

jrose-apple committed:
Add a RangeSpecificDiagnosticConsumer

If a top-level diagnostic is in a particular source range, the
RangeSpecificDiagnosticConsumer will funnel it and any attached notes
to a particular "sub-consumer" designated for that range (intended to
be used with whole files). If it's not in a range designated for any
sub-consumer, the diagnostic is passed to all registered
sub-consumers.

This is intended to be used for batch mode, so that diagnostics that
are definitely associated with a particular file can be emitted in
that file's .dia output, while diagnostics that may be associated with
other files (such as those that come from Clang) will still get
presented to the user.

jrose-apple commented:
@swift-ci Please test

graydon reviewed (approved):
Looks great!

jrose-apple commented:
@davidungar: Still waiting for another round of review, or you could jump to #14900 and see how it's used (and the unfortunate but unavoidable switch to basing on file names).

jrose-apple commented:
(We moved to just doing it all at once in #14900. Closing.)
